lib/rex/post/meterpreter/packet.rb

```diff
@@ -693,17 +819,31 @@ def to_r
   # passing it on to the default functionality that can parse
   # the TLV values.
   #
-  def from_r(bytes)
-    xor_key = bytes[0,4].unpack('N')[0]
-    super(xor_bytes(xor_key, bytes[4, bytes.length]))
+  def from_r(aes_key=nil)
+    xor_key = self.raw[XOR_KEY_OFF, XOR_KEY_SIZE]
+    data = xor_key + xor_bytes(xor_key, self.raw[SESSION_GUID_OFF, self.raw.length - SESSION_GUID_OFF])
+    self.session_guid = data[SESSION_GUID_OFF, SESSION_GUID_SIZE]
+    encrypted_flag = data[ENCRYPTED_FLAG_OFF, ENCRYPTED_FLAG_SIZE] == "\x01"
+    packet_length = data[PACKET_LENGTH_OFF, PACKET_LENGTH_SIZE]
+    packet_type = data[PACKET_TYPE_OFF, PACKET_TYPE_SIZE]
+    self.type = packet_type.unpack('N')[0]
+    if encrypted_flag
```

Comment:
Hmm. But a passive MitM who knows the session GUID can spoof TLV packets that say "is ok, this packet isn't encrypted so you have no way of knowing that it came from the real client, don't bother decrypting it. Here have a fistful of cleartext TLV's maybe one of them is relevant to you right now?" Which would violate the integrity of the session?

> We have to trust something in the protocol

Do we? Can't we track whether the session pertaining to a particular GUID is expected to be encrypted?

You know the architecture and the future plans, I'm just throwing ideas out there. I'm guessing there must be a reason why we *can't* go "Ah this GUID pertains to a session that I expect to get encrypted TLV's for. This packet has `!encrypted_flag` chuck it away."